api/src/api.c

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (c) 2016, ARM Limited, All Rights Reserved
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "api/inc/uvisor-lib.h"
+#include <stdint.h>
+
+
+/******************************** namespace ********************************/
+
+int uvisor_box_namespace(int box_id, char *box_namespace, size_t length)
+{
+    return uvisor_api.box_namespace(box_id, box_namespace, length);
+}
+
+/********************************** debug **********************************/
+void uvisor_debug_init(const TUvisorDebugDriver * const driver)
+{
+    uvisor_api.debug_init(driver);
+}
+
+void uvisor_error(THaltUserError reason)
+{
+    uvisor_api.error(reason);
+}
+
+/******************************** page heap ********************************/
+int uvisor_page_malloc(UvisorPageTable * const table)
+{
+    return uvisor_api.page.malloc(table);
```

Comment:
Why this level of indirection? There are 2 important design requirements which I'm not sure we can satisfy at the same time with this approach:

1. All entry points in one file.
2. The one file shows immediately which API will be run in P, NP, S, NS mode.

`1.` is satisfied here, but then the check on "where are you coming from" will be scattered throughout the other several files, compromising `2.`.

Also, we already discussed this: By using this level of indirection the boundary P/NP is shifted. It is more natural that the uvisor-lib runs the very last instruction that is executed in NP. The uvisor-core instead would run the very first instruction which is executed in P mode. In that sense the `svc` should be called from here (lib) and not from the core.